components/tikv_util/src/memory.rs

```diff
@@ -34,9 +35,68 @@ pub trait HeapSize {
     }
 }
 
-impl HeapSize for [u8] {
+macro_rules! impl_heap_size{
+    (
+        $($typ: ty,)+
+    ) => {
+        $(
+            impl HeapSize for $typ {
+                fn heap_size(&self) -> usize {
+                    std::mem::size_of::<Self>()
+                }
+            }
+        )+
+    }
+}
+
+impl_heap_size! {
+    // These types are not necessary stored in heap as they can be inlined
+    // in structures.
+    // TODO: We may need to add a new method to distinguish the difference.
+    u8, bool, u64,
+}
+
+impl<T: HeapSize> HeapSize for [T] {
+    fn heap_size(&self) -> usize {
+        if self.is_empty() {
+            0
+        } else {
+            self.len() * self[0].heap_size()
+        }
+    }
+}
+
+impl<T: HeapSize> HeapSize for Vec<T> {
+    fn heap_size(&self) -> usize {
+        // NB: It's an approximation of Vec<T> heap usage, because its capacity
+        // may be large then its length, and the unused space is also in heap.
+        self.as_slice().heap_size()
```

Comment:
It seems the inline part of T is ignored. The accurate heap size seem to be `self.capacity() * size_of::<T>()` + sum of heap sizes of the items.